### PR TITLE
Feature/main/jasonhalpern/migrate to sf

### PR DESCRIPTION
**Pull request by jasonhalpern**

### CODE REVIEW

**Comment by jasonhalpern** [comment]
Attempt to update to Salesforce/CLI from deprecated SFDX-cli (see: https://developer.salesforce.com/docs/atlas.en-us.sfdx_setup.meta/sfdx_setup/sfdx_setup_move_to_sf_v2.htm )

File: action.yml [hunk 1]
```diff
@@ -29,20 +29,31 @@ inputs:
     default: 'HEAD'
   api-version:
     description: 'SF API version to use during validate/deploy'
-    required: true
-    default: '51.0'
-  sfdx-download-url:
-    description: 'URL to download SFDX Linux client. Uses the XZ compressed version. Default is the stable version. Other version available at https://developer.salesforce.com/media/salesforce-cli/sfdx/versions/sfdx-linux-arm-tar-xz.json'
-    required: true
-    default: 'https://developer.salesforce.com/media/salesforce-cli/sfdx/channels/stable/sfdx-linux-x64.tar.xz'
+    required: false
```
Comment (ricka) [inline action.yml:32]: I suggest not doing this...pinning API versions just like we do with the package-lock.json is preferred here for me.

**Comment by jasonhalpern** [comment]
@ricka I removed the npm install, so now npm ci should install based on package-lock.json.  

I also added a conditional check to only install the sf plugin if it's missing (shaves 30 sec or so off of a run if so) and also a conditional check to force test classes to execute IF the test level wasn't specified (so sandbox will default to not running tests) but the code contains a change to a test class.  

If you think we should always leave this up to the caller (including if they want to be stupid and change a test class but not actually run it), I can revert this change... let me know your thoughts.

File: sfdx-cicd.sh [hunk 2]
```diff
@@ -1,54 +1,70 @@
...
   echo "Bad Validate param...choose true or false"
   exit 2
 fi
 
+# Set Test_Level (or use org default)
 TEST_LEVEL=`echo $TEST_LEVEL | tr '[:upper:]' '[:lower:]'`
+
+# Override sandbox default (notestrun) if TEST_LEVEL not specified
+# but package contains test class changes
+if [[ -n $(find diffdeploy/ -name '*test.cls') ]] && [[ $TEST_LEVEL = '' ]]; then
```
Comment (ricka) [inline sfdx-cicd.sh:39]: I think this is trying to be a little too clever, and also isn't correct.

First, with this approach, you'd have to use `-iname` for it to stand a chance. Second, plenty of test classes in plenty of orgs don't end in the word `test`. Our own org has quite a few. Sometimes the start with test, sometimes they end with it, sometimes they omit it all together.

If you want to go down that route, you really need to be grepping for `@istest` case insensitive within the body of the file. Even that might not be 100% in the case of comments or other edge cases I haven't thought about, but it will avoid issues with either people not following a convention or not having a convention.

The problem though this only runs the tests when the body of the test has changed. What you really want to do is test code that has changed by test classes that cover it, as historically test cases do not change very often. This is the problem we started to solve at RH but they didn't allow us to invest enough time to solve it (happy to talk about the solution over chat).

In short, I would just pull this out and make people be explicit about it vs trying to assume what they meant for them.

**Comment by jasonhalpern** [comment]
@ricka Removed the test class detection / Test Level defaulting per our conversation, caller can determine the level they want to run.  If this looks good, could you merge and maybe give it a v2 release so I can test it out?

File: sfdx-cicd.sh [hunk 3]
```diff
@@ -1,54 +1,63 @@
...
-# Auth sfdx into org with auth url
-echo $SFDX_AUTH_URL > sfdx_auth.txt
-sfdx force:auth:sfdxurl:store -f sfdx_auth.txt -s -a SFOrg
-rm sfdx_auth.txt
-
-echo "sfdx diff command: sfdx sfpowerkit:project:diff -r $REVISION_FROM -t $REVISION_TO -d diffdeploy"
-# Prepare diff
-sfdx sfpowerkit:project:diff -r $REVISION_FROM -t $REVISION_TO -d diffdeploy
+# Check if API_VERSION was specified
+if [[ $API_VERSION != '' ]]; then
```
Comment (ricka) [inline sfdx-cicd.sh:57]: This won't work for cases that API_VERSION isn't specified.

**Review by ricka — CHANGES_REQUESTED** [review]

**Comment by jasonhalpern** [comment]
@ricka Updated the usage text to clearly explain order of how API-VERSION is determined by the SFDX CLI (see: https://developer.salesforce.com/docs/atlas.en-us.sfdx_setup.meta/sfdx_setup/sfdx_setup_apiversion.htm ). 

Value in sfdx-project.json will OVERRIDE any value passed on the command line.

So the API-VERSION flag remains _optional_. If it is specified, AND the sfdx-project.json file has no value for sourceApiVersion, it will be used. 

If it is NOT specified, **and** there is no sourceApiVersion value in sfdx-project.json, it will use the current org default API version and throw a warning message to the log.

**Comment by ricka** [comment]
@jasonhalpern I think this is good for now. Probably better to have code to remove the `sfdx-project.json` before deploy or update the file's contents so we can force sf cli to use the version we specify long term. I'll add an issue for that.

Crazy to me to think that sf will ignore a cli flag in favor of the contents of a file.

**Review by ricka — APPROVED** [review]
👍
